src/modules/ssh_docker/mappers/docker_container.ts

```diff
@@ -0,0 +1,134 @@
+      _.isEqual(a.networkSettings, b.networkSettings) &&
+      _.isEqual(a.mounts, b.mounts);
+    return res;
+  };
+
+  private containerMapper(serverName: string, container: ContainerInfo): DockerContainer {
+    const out = new DockerContainer();
+    out.serverName = serverName;
+    out.containerId = container.Id;
+    out.name = container.Names[0];
+    out.image = container.Image;
+    out.imageId = container.ImageID;
+    out.command = container.Command;
+    out.created = new Date(container.Created * 1000);
+    out.ports = container.Ports;
+    out.labels = container.Labels;
+    out.state = container.State;
+    out.status = container.Status;
+    out.hostConfig = container.HostConfig;
+    out.networkSettings = container.NetworkSettings;
+    out.mounts = container.Mounts;
+    return out;
+  }
+
+  private async getDockerForServer(ctx: Context, serverName: string): Promise<Docker | undefined> {
+    const sshClient: SSH2Promise = await ctx.getSshClient(serverName);
+    const sshConfig = sshClient.config[0];
+    const modem = new Modem({
+      protocol: 'ssh',
+      host: sshConfig.host,
+      port: sshConfig.port,
+      username: sshConfig.username,
+      sshOptions: {
+        privateKey: sshConfig.privateKey,
+        passphrase: sshConfig.passphrase,
+      },
+    });
+    const docker = new Docker({ modem } as DockerOptions);
+    try {
+      await docker.ping();
+    } catch (e) {
+      return undefined;
+    }
+    return docker;
+  }
+
+  cloud = new Crud({
+    create: async (es: DockerContainer[], ctx: Context) => {
+      // create docker containers through RPC
+      await this.db.delete(es, ctx);
+    },
+    read: async (ctx: Context, id?: string) => {
+      if (!!id) {
+        const { serverName, containerId } = this.idFields(id);
+        const docker = await this.getDockerForServer(ctx, serverName);
+        if (!docker) return;
+
+        const container = (await docker.listContainers()).find(c => c.Id === containerId);
+        if (!container) return undefined;
+        return this.containerMapper(serverName, container);
+      } else {
+        const out: DockerContainer[] = [];
+        const sshCredentials: SshCredentials[] = await sshAccounts.sshCredentials.cloud.read(ctx);
+        const serverNames = sshCredentials.map(c => c.name);
+        for (const serverName of serverNames) {
+          const docker = await this.getDockerForServer(ctx, serverName);
+          if (!docker) continue;
+
+          const containers = await docker.listContainers({ all: true });
+          for (const rawContainer of containers) {
+            const container = this.containerMapper(serverName, rawContainer);
+            out.push(container);
+          }
+        }
+        return out;
+      }
+    },
+    update: async (es: DockerContainer[], ctx: Context) => {
+      const out = [];
+      for (const e of es) {
+        // restore because you can't change a container
```

Comment:
I think the `status` column should be mutable so you can change the state of a container to `stopped` or `started`.

Also, it would be *better* if instead of undoing the changes for the other columns, it flags this update as a `replace` and does a delete/create flow.